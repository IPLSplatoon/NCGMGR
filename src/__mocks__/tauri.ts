export const mockTauriEvent = {
    listen: jest.fn()
}

jest.mock('@tauri-apps/api/event', () => mockTauriEvent)

export const mockTauri = {
    invoke: jest.fn()
}

jest.mock('@tauri-apps/api/tauri', () => mockTauri)

export const mockTauriDialog = {
    open: jest.fn()
}

jest.mock('@tauri-apps/api/dialog', () => mockTauriDialog)

export const mockTauriFs = {
    readDir: jest.fn(),
    readTextFile: jest.fn(),
    removeDir: jest.fn(),
    removeFile: jest.fn(),
    createDir: jest.fn(),
    writeFile: jest.fn()
}

jest.mock('@tauri-apps/api/fs', () => mockTauriFs)

export const mockTauriOs = {
    type: jest.fn()
}

jest.mock('@tauri-apps/api/os', () => mockTauriOs)

export const mockTauriShell = {
    open: jest.fn()
}

jest.mock('@tauri-apps/api/shell', () => mockTauriShell)
