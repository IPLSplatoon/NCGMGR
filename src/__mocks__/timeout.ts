export function mockTimeout (): void {
    jest.spyOn(window, 'setTimeout').mockImplementation(handler => {
        // noinspection SuspiciousTypeOfGuard
        if (typeof handler !== 'string') {
            handler()
        }
        return 0
    })
}
