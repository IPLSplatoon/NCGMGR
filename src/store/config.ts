import { createStore, Store, useStore } from 'vuex'
import { InjectionKey } from 'vue'

export const LOCAL_STORAGE_CONFIG_KEY = 'config'

export interface Configuration {
    installPath: string
}

export const configStore = createStore<Configuration>({
    state: {
        installPath: ''
    },
    mutations: {
        setInstallPath: (state, installPath: string) => {
            state.installPath = installPath
        },
        setState: (state, config: Configuration) => {
            Object.assign(state, config)
        }
    },
    actions: {
        load: store => {
            const config = JSON.parse(localStorage.getItem(LOCAL_STORAGE_CONFIG_KEY) ?? '{}')
            store.commit('setState', {
                ...store.state,
                ...config
            })
        },
        persist: store => {
            localStorage.setItem(LOCAL_STORAGE_CONFIG_KEY, JSON.stringify(store.state))
        }
    }
})

export const configStoreKey: InjectionKey<Store<Configuration>> = Symbol()

export function useConfigStore (): Store<Configuration> {
    return useStore(configStoreKey)
}
