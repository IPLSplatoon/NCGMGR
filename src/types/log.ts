export interface LogEvent {
    message: string
    type?: 'success' | 'info' | 'error'
}

export interface ProgressEvent {
    step: number
    // eslint-disable-next-line camelcase
    max_step: number
}

export enum ActionState {
    COMPLETED_SUCCESS,
    COMPLETED_ERROR,
    INCOMPLETE
}
