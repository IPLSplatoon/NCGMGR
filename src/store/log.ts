import { LogEvent } from '@/types/log'
import { createStore, Store, useStore } from 'vuex'
import { InjectionKey } from 'vue'
import { Event, listen, UnlistenFn } from '@tauri-apps/api/event'

export interface LogStore {
    lines: Record<string, LogEvent[]>
    completed: Record<string, boolean>
}

const unlistenFns: Record<string, UnlistenFn> = {}

export const logStore = createStore<LogStore>({
    state: {
        lines: {},
        completed: {}
    },
    mutations: {
        insertLine (state, { line, key }: { line: LogEvent, key: string }) {
            if (!state.lines[key]) {
                state.lines[key] = []
            }
            state.lines[key].push(line)
        },
        reset (state, key: string) {
            state.lines[key] = []
            state.completed[key] = false
        },
        setCompleted (state, { key, completed }: { key: string, completed: boolean }) {
            state.completed[key] = completed
        }
    },
    actions: {
        listen (store, key: string) {
            if (unlistenFns[key]) {
                return
            }
            return listen(`log:${key}`, (event: Event<LogEvent>) => {
                store.commit('insertLine', { line: event.payload, key })
            }).then(unlisten => {
                unlistenFns[key] = unlisten
            })
        },
        unlisten (store, key: string) {
            if (unlistenFns[key]) {
                unlistenFns[key]()
            }
        },
        logPromiseResult (store, { promise, key, noLogOnSuccess }: { promise: Promise<unknown>, key: string, noLogOnSuccess: boolean }) {
            promise.then(() => {
                if (!noLogOnSuccess) {
                    store.commit('insertLine', {
                        line: {
                            message: 'Success!',
                            type: 'success'
                        },
                        key
                    })
                }
            }).catch(e => {
                store.commit('insertLine', {
                    line: {
                        message: String(e),
                        type: 'error'
                    },
                    key
                })
            }).finally(() => {
                store.commit('setCompleted', { key, completed: true })
            })
        }
    }
})

export const logStoreKey: InjectionKey<Store<LogStore>> = Symbol()

export function useLogStore (): Store<LogStore> {
    return useStore(logStoreKey)
}
