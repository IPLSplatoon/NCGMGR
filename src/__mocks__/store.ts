import { createStore, Store } from 'vuex'
import { Configuration } from '@/store/config'
import { LogStore } from '@/store/log'

export function createConfigStore (): Store<Configuration> {
    return createStore<Configuration>({
        state: {
            installPath: '/install/path'
        },
        actions: {
            load: jest.fn(),
            persist: jest.fn()
        },
        mutations: {
            setInstallPath: jest.fn()
        }
    })
}

export function createLogStore (): Store<LogStore> {
    return createStore<LogStore>({
        state: {
            lines: [],
            completed: false
        },
        mutations: {
            reset: jest.fn(),
            insertLine: jest.fn(),
            setCompleted: jest.fn()
        }
    })
}
