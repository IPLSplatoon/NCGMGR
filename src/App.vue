<template>
    <install-manager />
    <bundle-manager v-if="nodecgInstalled" class="m-t-8" />
</template>

<script lang="ts">
import { computed, defineComponent } from 'vue'
import InstallManager from '@/components/installManager.vue'
import BundleManager from '@/components/bundleManager.vue'
import { InstallStatus, useNodecgStore } from '@/store/nodecg'

export default defineComponent({
    name: 'App',

    components: { BundleManager, InstallManager },

    setup () {
        const nodecgStore = useNodecgStore()

        return {
            nodecgInstalled: computed(() => nodecgStore.state.status.installStatus === InstallStatus.INSTALLED)
        }
    }
})
</script>
