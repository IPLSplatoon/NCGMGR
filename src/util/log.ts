import { logStore } from '@/store/log'

export function logPromiseResult (promise: Promise<unknown>): void {
    promise.then(() => {
        logStore.commit('insertLine', {
            message: 'Success!',
            type: 'success'
        })
    }).catch(e => {
        logStore.commit('insertLine', {
            message: String(e),
            type: 'error'
        })
    }).finally(() => {
        logStore.commit('setCompleted', true)
    })
}
