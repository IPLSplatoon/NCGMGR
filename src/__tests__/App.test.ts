import { mockTauri, mockTauriDialog } from '@/__mocks__/tauri'
import App from '../App.vue'
import { shallowMount } from '@vue/test-utils'
import { configStoreKey } from '@/store/config'
import { createConfigStore, createLogStore } from '@/__mocks__/store'
import { logStoreKey } from '@/store/log'

describe('App', () => {
    it('matches snapshot', () => {
        const wrapper = shallowMount(App, {
            global: {
                plugins: [
                    [createConfigStore(), configStoreKey]
                ]
            }
        })

        expect(wrapper.html()).toMatchSnapshot()
    })

    it('handles directory selection', async () => {
        const store = createConfigStore()
        jest.spyOn(store, 'dispatch')
        jest.spyOn(store, 'commit')
        const wrapper = shallowMount(App, {
            global: {
                plugins: [
                    [store, configStoreKey]
                ]
            }
        })
        mockTauriDialog.open.mockResolvedValue('/new/path')

        await wrapper.get('[data-test="install-directory-select-button"]').trigger('click')

        expect(store.commit).toHaveBeenCalledWith('setInstallPath', '/new/path')
        expect(store.dispatch).toHaveBeenCalledWith('persist')
        expect(store.dispatch).toHaveBeenCalledTimes(2)
    })

    it('handles directory selection getting cancelled', async () => {
        const store = createConfigStore()
        jest.spyOn(store, 'dispatch')
        jest.spyOn(store, 'commit')
        const wrapper = shallowMount(App, {
            global: {
                plugins: [
                    [store, configStoreKey]
                ]
            }
        })
        mockTauriDialog.open.mockResolvedValue(null)

        await wrapper.get('[data-test="install-directory-select-button"]').trigger('click')

        expect(store.commit).not.toHaveBeenCalled()
        expect(store.dispatch).toHaveBeenCalledTimes(1)
    })

    it('disables installation if no install directory is selected', () => {
        const store = createConfigStore()
        store.state.installPath = ''
        const wrapper = shallowMount(App, {
            global: {
                plugins: [
                    [store, configStoreKey]
                ]
            }
        })

        expect((wrapper.get('[data-test="install-button"]').element as HTMLButtonElement).disabled).toEqual(true)
    })

    it('enables installation if install directory is selected', () => {
        const store = createConfigStore()
        store.state.installPath = '/install/path'
        const wrapper = shallowMount(App, {
            global: {
                plugins: [
                    [store, configStoreKey]
                ]
            }
        })

        expect((wrapper.get('[data-test="install-button"]').element as HTMLButtonElement).disabled).toEqual(false)
    })

    it('handles installation', () => {
        const configStore = createConfigStore()
        configStore.state.installPath = '/install/path'
        const logStore = createLogStore()
        jest.spyOn(logStore, 'commit')
        mockTauri.invoke.mockResolvedValue({})
        const wrapper = shallowMount(App, {
            global: {
                plugins: [
                    [configStore, configStoreKey],
                    [logStore, logStoreKey]
                ]
            }
        })

        wrapper.get('[data-test="install-button"]').trigger('click')

        expect(logStore.commit).toHaveBeenCalledWith('reset')
        expect(mockTauri.invoke).toHaveBeenCalledWith('install_nodecg', { path: '/install/path' })
    })
})
