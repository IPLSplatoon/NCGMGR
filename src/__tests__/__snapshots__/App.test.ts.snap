// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches snapshot 1`] = `
<div class="layout vertical">
  <div>Installation folder: "/install/path"</div><button data-test="install-directory-select-button">Select folder</button><button data-test="install-button">Install</button>
</div>
<log-overlay-stub title="Installing..." visible="false"></log-overlay-stub>
`;
